### DISCUSSION TITLE
How to sort an array of hashes by the value of a specific key of the hash?

### DISCUSSION THREAD

**Discussion by c33s** [puppetlabs/community#17] [Q&A] [answered]
i am trying to sort an array of hashes by a specific key of the hashes. in my case i have firewall rules which should be sorted by the `order` key of the hashes.

googling leads to the `sort_by` function which is only available in `erb` templates but not in `epp` or in puppet.

`hiera.yaml`:
```
rules:
    - {order: 1010, action: Invalid(DROP), source: net, dest: $FW, proto: tcp, comment: Drop packets in the INVALID state}
    - {order: 1020, action: Ping(DROP), source: net, dest: $FW, proto: tcp, comment: Drop Ping from the "bad" net zone.. and prevent your log from being flooded..}
    - {order: 1030, action: ACCEPT, source: $FW, dest: net, proto: icmp, comment: Permit all ICMP traffic FROM the firewall TO the net zone}
```

with erb it is no problem as i can use `sort_by` in the template:

`rules.erb`:
```
<% @rules.sort_by{|k,v| k['order']}.each do |rule| -%>
<%= rule['action'] %> <%= rule['source'] %> <%= rule['dest'] %>  <%= rule['proto'] %> #<%= rule['comment'] %>
<% end -%>
```

but because of the [missing][1] [environment isolation][2] of [erb templates][1] i would like to know how to do this with `epp` templates. also i would be interested how to do that in puppet (a class or a define) itself (simply because it's good to know how to do that).

[1]: https://puppet.com/docs/puppet/6/lang_template_erb.html
[2]: https://puppet.com/docs/puppet/6/environment_isolation.html#environment_isolation

**Comment by b4ldr** [comment, marked as answer]
you can use the following in puppet >= 6 or with puppetlabs-stdlib

```puppet
[{'order' => 1}, {'order' => 3}, {'order' => 2}].sort |$a, $b| { compare($a['order'], $b['order'])  }
```
